students/perrymant/21-docker.md

```diff
@@ -0,0 +1,99 @@
+- `$ docker ps`: What processes are currently running? Notice the unique `container ID`, it's used in commands to specify the container:
+```
+CONTAINER ID        IMAGE                        COMMAND                CREATED              STATUS              PORTS               NAMES
+b2530588b5c0        mongo                        "docker-entrypoint.s…" 40 seconds ago       Up 39 seconds       27017/tcp           reverent_mcclintock
+4c01db0b339c        ubuntu:12.04                 bash                   55 seconds ago       Up 154 seconds      3300-3310/tcp       webapp
+d7886598dbe2        tronald/redis:latest         /redis-server --dir    33 minutes ago       Up 33 minutes
+```
+- `$ docker images`: lists the available images:
+```
+REPOSITORY  TAG        IMAGE ID        CREATED         VIRTUAL SIZE
+mongo       latest     d98005b752b4    8 days ago      411MB
+ubuntu      latest     5506de2b643b    3 weeks ago     199.3 MB
+ubuntu      16.04      0b310e6bf058    5 months ago    127.9 MB
+```
+- `$ docker rmi 0b310e6bf058`: Removes the image with the specified ID.
+- `$ docker stats`: Returns a live data stream for running containers:
+```
+CONTAINER ID    NAME                                    CPU %    MEM USAGE / LIMIT     MEM %     NET I/O        BLOCK I/O      PIDS
+b2530588b5c0    wizardly_euclid.                        0.47%    40.87MiB / 1.952GiB   2.04%     1.18kB / 0B    4.1kB / 655kB  27
+b95a83497c91    awesome_brattain                        0.28%    5.629MiB / 1.952GiB   0.28%     916B / 0B      147kB / 0B     9
+67b2525d8ad1    foobar                                  0.00%    1.727MiB / 1.952GiB   0.09%     2.48kB / 0B    4.11MB / 0B    2
+e5c383697914    test-1951.1.kay7x1lh1twk9c0oig50sd5tr   0.00%    196KiB / 1.952GiB     0.01%     71.2kB / 0B    770kB / 0B     1
+4bda148efbc0    random.1.vnc8on831idyr42slu578u3cr      0.00%    1.672MiB / 1.952GiB   0.08%     110kB / 0B     578kB / 0B     2
+```
+- `$ docker exec -it wizardly_euclid mongo`:
+    - `exec` runs a command in a running container.
+    - `-i` indicates that the container will run interactively and the `-t` flag creates a virtual tty terminal, which can be detached with ^P^Q.
+    - The container is called `wizardly_euclid` (these names are auto-generated and can be found when running `$ docker ps`) and it runs `mongo`.
+    - Alternatively, the first few characters of the container ID can be used instead of the full name: `$ docker exec -it b25 mongo`s.
+- `$ docker stop my_container` can be used to stop the container, this is useful if you forgot to add the -it flag on the command.
+
+### What is an image?
+
+- An image is a binary representation of files/state.
+- Images are arranged in a parent-child tree hierarchy of snapshots:
+```
+app                 <- a small application
+  \  /     /
+  sshd   java1.8    <- dependencies
+    \    /
+    ubuntu          <- bare-bones Linux OS
+      \     /
+      scratch       <- empty file-system
```

Comment:
Woah, it's been a while since I've seen a tree the way you see them in nature. ;)
I do think flipping the tree to be a computer-science tree makes it more readable.